designs/aws/aws-launch-templates-v2.md

```diff
@@ -0,0 +1,146 @@
+## Karpenter and AWS LaunchTemplates
+
+
+This document focuses on how to evolve Karpenter’s support for AWS LaunchTemplates.
+
+### **What is the problem?**
+
+Karpenter currently lets customers control several properties of the worker node that are being provisioned via the Provisioner. While this includes most core, required parameters like the instanceType, subnets, securityGroups, not all EC2 level parameters are supported.
+
+If additional configuration is required (say to use a specific AMI), we ask that a complete LaunchTemplate is provided to the Karpenter provisioner. This route has had a few advantages -
+
+1. Users can fully customize their worker node and use any available EC2 feature that’s exposed via launch templates.
+2. Partially filled out launch templates are difficult to interpret. Anyone specifying a custom AMI to use will also need to explicitly specify how it needs to be bootstrapped via UserData. This is important because Karpenter cannot make guesses on how to start the kubelet on an unknown AMI.
+
+The main, critical downside of this approach has been that a user needs to maintain a fully-formed launch template even when they’re only looking to customize a single property of their worker nodes. Consider these examples -
+
+1. In order to override [the root volume](https://github.com/aws/karpenter/issues/885), you’d need to author an entire launch template containing security groups, userData etc.
+2. In order to just have all statically provisioned [volumes be encrypted](https://github.com/aws/karpenter/issues/933), you need to author an entire launch template.
+
+While some users do like complete control over the launch template since they customize everything (kOps), most only need to override a few settings and would therefore like those to be present as knobs on the provisioner itself, and for all other settings to be figured out by Karpenter.
```

Comment:
The challenge here is really that karpenter assumes EKS and anything that is not EKS need different launch templates. Rancher, OKD etc all will have this requirement. Had Karpenter been made to work firstly with kOps, EKS would have this requirement.

I don't think it is feasible that Karpenter truly can manage LTs that will work across all these installers, and appreciate that for EKS this is needed since there is no tooling for managing LTs. But other installers do provide that tooling. So certainly no kOps _user_ will _want_ to manage their own LTs. In fact, kOps do not allow custom LTs, and I haven't seen a single request for such a feature (our approach is somewhat similar to approach 2 below). End goal being sort of the same as this aspect of Karpenter: hide the complexity of LTs.

Reply:
This suggests to me that we probably need to `provider: Kops` or `provider: AWSKops` implementation, since they vary so much.